### ISSUE TITLE
Generic Bessel functions

### ISSUE THREAD

**Issue by 0x00b1** [pytorch/pytorch#76324]
### 🚀 The feature, motivation and pitch

I have a pull request open that adds the remaining so-called fast Bessel functions (e.g., `j0`, `y0`, `k0`, etc.) but I now want to start thinking about adding the generic Bessel functions as they are extremely useful (and are needed to provided nice derivatives for the existing and previously proposed Bessel functions). This is a sketch of the API I have been thinking about:

`torch.special.j(x: Tensor, order) → Tensor`, Bessel function of `order`.
`torch.special.je(x: Tensor, order) → Tensor`, exponentially scaled Bessel function of `order`.
`torch.special.y(x: Tensor, order) → Tensor`, Bessel function of the second kind of `order`.
`torch.special.ye(x: Tensor, order) → Tensor`, exponentially scaled Bessel function of the second kind of `order`.
`torch.special.k(x: Tensor, order) → Tensor`, modified Bessel function of the second kind of `order`.
`torch.special.ke(x: Tensor, order) → Tensor`, exponentially scaled modified Bessel function of the second kind of `order`.
`torch.special.i(x: Tensor, order) → Tensor`, modified Bessel function of the first kind of `order`.
`torch.special.ie(x: Tensor, order) → Tensor`, exponentially scaled modified Bessel function of the first kind of `order`.
`torch.special.h1(x: Tensor, order) → Tensor`, Hankel function of the first kind of `order`.
`torch.special.h1e(x: Tensor, order) → Tensor`, exponentially scaled Hankel function of the first kind of `order`.
`torch.special.h2(x: Tensor, order) → Tensor`, Hankel function of the second kind of `order`.
`torch.special.h2e(x: Tensor, order) → Tensor`, exponentially scaled Hankel function of the second kind of `order`.

**Comment by vadimkantorov** [comment]
Some naming considerations:
- it may be simpler to google things if they have some prefix
- it may also save some code when importing just these functions from the module `from torch.special import k` as there currently may be clashes with other variables `k`/`i` (`as bessel_k` is possible of course, but doing without this and keeping the naming standard across codebases may be good)
- on the other side there are benefits to keep the same naming as in other existing frameworks: scipy / matlab etc...

**Comment by mruberry** [comment]
cc @rgommers @Lezcano

**Comment by mruberry** [comment]
When it comes to operator naming we can have aliases, too, if there are two reasonable names for operations.

@0x00b1 -- how would this interface compare or be consistent with PyTorch's current operations and SciPy's special operations? Are there other popular frameworks or libraries we can look at for datapoints, too, like Matlab or Julia?

**Comment by rgommers** [comment]
For SciPy [docs](http://scipy.github.io/devdocs/reference/special.html#bessel-functions) most functions have an extra "v", so `jv`, `jve` instead of `j`, `je` as proposed here. And `hankel` is spelled in full.

For Matlab I can only find a subset [here](https://www.mathworks.com/help/matlab/special-functions.html?s_tid=CRUX_lftnav) so quickly, but names are `besselj` etc.

One-letter function names are an anti-pattern imho. There's no good reason not to prepend `bessel_` and `hankel_` here.

**Comment by Lezcano** [comment]
Disclaimer: I don't know much more about Bessel functions than what you study on a first / second course on ODEs (which is not much).

Now, my 2cts on this are:

I would go with `bessel_*` names rather than just with the names from `scipy`. The ideas in https://github.com/pytorch/pytorch/issues/76324#issuecomment-1109014097 for doing this are very compelling. Even more, otherwise you get awkward things like scipy's [`lmbda` function](https://docs.scipy.org/doc/scipy/reference/generated/scipy.special.lmbda.html#scipy.special.lmbda) (I know, right?).

It's not clear a priori to me:
- How difficult is it to implement these functions on CUDA.
- How difficult is it to implement their derivatives: Can we just differentiate the forward? Are their derivatives implemented in terms of other Bessel functions?
- How difficult is to implement these functions for complex inputs: We currently do not support complex inputs in `torch.special` (a pity), but perhaps looking forward it could be interesting to see whether we can implement real and complex support for the new functions from the get go.

@0x00b1 how useful are these functions if they just take real inputs?

**Comment by 0x00b1** [comment]
> How difficult is it to implement these functions on CUDA.

Easy.

> How difficult is it to implement their derivatives: Can we just differentiate the forward? Are their derivatives implemented in terms of other Bessel functions?

We cannot just differentiate the forward but they are relatively easy to implement and can probably be implemented in their entirety in derivatives.yaml as they are usually implemented in terms of other Bessel functions.

> How difficult is to implement these functions for complex inputs: We currently do not support complex inputs in torch.special (a pity), but perhaps looking forward it could be interesting to see whether we can implement real and complex support for the new functions from the get go.

Not easy but not especially challenging. I agree that complex inputs would be super valuable. In fact, this is my use case!

**Comment by 0x00b1** [comment]
> One-letter function names are an anti-pattern imho. There's no good reason not to prepend bessel_ and hankel_ here.

I agree. I also like the written format for special functions as it saves me a trip to Wikipedia to remember the notation. However, one challenge is consistency with `i0`, i0e`, i1`, and `i1e`. What would you do here?

**Comment by rgommers** [comment]
> However, one challenge is consistency with `i0`, i0e`, i1`, and `i1e`. What would you do here?

The `torch.special` versions of those functions are pretty new and not heavily used yet I assume. So I would just give them better/consistent new names and deprecate the existing ones. 

I'm not sure if that was considered when they were first implemented. @kshitij12345 can you comment on that?

**Comment by kshitij12345** [comment]
While adding these functions, the idea was the match the names from `scipy.special` as `torch.special` is meant to be analogous. However I agree that one-letter names are not the must intuitive and it would be nice to have clear prefix.

Like @vadimkantorov mentioned we can add these new aliases and gradually deprecate the shorter names (or keep them but publicize the new names).

**Comment by hchau630** [comment]
would it also be possible to add the general bessel functions with real order and complex arguments (`jv`, `jve`, `yv`, `yve`, `iv`, `ive`, `kv`, `kve`)? Personally I've encountered the need for complex arguments for `k0`.

**Comment by Marius-Juston** [comment]
Hello, does anyone know if there has been progress on the addition of the generalized Modified Bessel function of the second kind $K_n(x)$? The `modified_bessel_k(input: Tensor, n: int = 0) → Tensor`  function. While I do not currently need it to be differentiable it would be nice to have this implemented.

**Comment by hchau630** [comment]
@Marius-Juston likely not. But if you don't need to handle complex number inputs then you can easily implement it yourself using the recurrence relations and the fact that k0 and k1 are already implemented.

**Comment by 0x00b1** [comment]
I should finish this!
